src/main/java/javaslang/control/Validation.java

```diff
@@ -0,0 +1,253 @@
+    }
+
+    static <E,T> Validation<E,T> failure(Supplier<E> supplier) {
+        return new Failure<>(supplier.get());
+    }
+
+    boolean isSuccess();
+
+    boolean isFailure();
+
+    Object get();
+
+    void foreach(Consumer<T> f);
+
+    @Override
+    boolean equals(Object o);
+
+    @Override
+    int hashCode();
+
+    @Override
+    String toString();
+
+    default <U> U fold(Function<? super E,? extends U> fail, Function<? super T,? extends U> success) {
+        if(isFailure()) {
+            E v = ((Failure<E,T>) this).get();
+            return fail.apply(v);
+        } else {
+            T v = ((Success<E,T>) this).get();
+            return success.apply(v);
+        }
+    }
+
+    default Validation<T,E> swap() {
+        if(isFailure()) {
+            E v = ((Failure<E,T>) this).get();
+            return Validation.success(v);
+        } else {
+            T v = ((Success<E,T>) this).get();
+            return Validation.failure(v);
+        }
+    }
+
+    @Override
+    default <U> Validation<E,U> map(Function<? super T,? extends U> mapper) {
+        if(isFailure()) {
+            E v = ((Failure<E,T>) this).get();
+            return Validation.failure(v);
+        } else {
+            T v = ((Success<E,T>) this).get();
+            return Validation.success(mapper.apply(v));
+        }
+    }
+
+    default <U,R> Validation<U,R> bimap(Function<? super E,? extends U> fail, Function<? super T,? extends R> success) {
+        if(isFailure()) {
+            E v = ((Failure<E,T>) this).get();
+            return Validation.failure(fail.apply(v));
+        } else {
+            T v = ((Success<E,T>) this).get();
+            return Validation.success(success.apply(v));
+        }
+    }
+
+    default <U> Validation<U,T> leftMap(Function<? super E,? extends U> f) {
+        if(isFailure()) {
+            E v = ((Failure<E,T>) this).get();
+            return Validation.failure(f.apply(v));
+        } else {
+            T v = ((Success<E,T>) this).get();
+            return Validation.success(v);
+        }
+    }
+
+    default <U> Validation<List<E>,U> ap(Validation<List<E>,Function1<T,U>> vv) {
+        if(isSuccess() && vv.isSuccess()) {
+            return success(() -> {
+                Function<T,U> f = ((Success<List<E>,Function1<T,U>>) vv).get();
+                return f.apply(((Success<E,T>)this).get());
+            });
+        } else if(isSuccess() && vv.isFailure()) {
+            return (Failure<List<E>,U>) vv;
+        } else if(isFailure() && vv.isSuccess()) {
+            E e = ((Failure<E,T>) this).get();
+            return failure(List.of(e));
+        } else {
+            return failure(() -> {
+                List<E> e  = ((Failure<List<E>,Function1<T,U>>) vv).get();
+                E  e2 = ((Failure<E,T>) this).get();
+                return e.append(e2);
+            });
+        }
+    }
+
+    default <U> ValidationBuilder<E,T,U> bld(Validation<E,U> v) {
```

Comment:
Without having dived into it, my feeling says that maybe there is a way to compose Applicatives without the need for builders... Need to think about it...